home-manager/home-manager

```diff
@@ -74,7 +74,20 @@ function setHomeManagerNixPath() {
     done
 }
 
+# get activationPackage of homeConfig reference, defaults to $USER if no attr is spedified
+function setFlakeAttribute() {
+    case $1 in
+        *#*) local name="${1#*#}";;
+        *)   local name="$USER";;
+    esac
+    export FLAKE_CONFIG_URI="${1%#*}#homeConfigurations.${name}"
```

Comment:
How difficult would it be to parameterize this across both `hostname` and `user`, as the [original nixflk implementation](https://github.com/divnix/devos/pull/93) did? I'd guess difficult since resolving the second `.` without some escape syntax is hard?

Generally, I think distinguishing based on host is more useful than username, since the differences will be things like OS or purpose, rather than just the user; I might even have different usernames across machines, because some are pre-configured for me and I have no choice of username.

Also, when running `home-manager --flake switch`, I'm unlikely to be managing multiple user accounts, since I will already be logged into a specific user on a machine and am deliberately installing from a particular git repo; I think different users are likely to have different repositories entirely.

Reply:
I mainly chose username, because it seemed like the natural analog to the hostname. It's of course only a suggestion, you can always build another "profile" with `home-manager switch --flake uri#<other-name>`. It wouldn't necessarily be too difficult to implement something like your suggestion, but I'm not sure if it's a good idea too tie home manager configurations to a machine. Then again, I'm not currently managing multiple accounts on multiple machines.
As a side note, you have to specify a flake uri in this version, otherwise argument parsing becomes complicated (and ambiguous).
Should: `home-manager switch --flake --commit-lock-file` build the flake in the registry called `--commit-lock-file` or the flake at the "default location" (`~/.config/nix/flake.nix`?) and commit the lock file?

Reply:
Fairly certain `nixos-rebuild` also requires the flake to be specified, I left the argument empty because the exact command isn't really relevant and there's too many URIs to consider for a nice general one ;)

But anyway, my point is; yeah, the username is a logical equivalent, but I'm not sure it's the best UI. From looking through GitHub repos, most people seem to parametrize based on hostname, not username, and multiple users in one config is an exception rather than the norm (nixflk supports it, but no forks seem to use it).

So I _fear_ it'll just be left empty and unused 99% of the time, while everyone ends up implementing their own host name matching. I'd like to avoid that scenario.

I have of course not seen all configs, just raising the use cases I can see.

Reply:
Just to add; not "empty and unused" as in "that part of the URI will always be empty", but "people will always need to specify their host name and the default will throw a bad error or deploy something entirely wrong".

Reply:
Considering how Home Manager is centered around the user and used outside of NixOS, I believe the username is the better choice.

It would also work with `sudo -u some-user home-manager switch --flake .`

Reply:
I agree that the user seems like a more expected target. Personally I created a script that supported both. It checks for a config named user@host, if it doesn't exist it falls back to just user. Not sure if that is the best way, but that ended up working well for me.

Reply:
BTW, from my own learnings from my own script I found you need to do this to support usernames containing `.`.


```
export FLAKE_CONFIG_URI="${1%#*}#homeConfigurations.\"${name}\""
```

I was happy that nix accepted the quotes, because before I found that I didn't find a way to avoid having the dots evaluated as an attribute accessor.

Reply:
Oh, didn't even consider that names could contain dots, good catch :+1:.
After some thinking I really like the @ idea, seems like the most natural way to allow both name-only and name+hostname configuratons. Would that work for you @TLATER?
Also, can you point me to where you've implemented it? It shouldn't be too difficult, just to double check.

Reply:
Oh, didn't even consider that names could contain dots, good catch :+1:.
After some thinking I really like the @ idea, seems like the most natural way to allow both name-only and name+hostname configuratons. Would that work for you @TLATER?
Also, can you point me to where you've implemented it? It shouldn't be too difficult, just to double check.

Reply:
Yep, this sounds good :)

Reply:
I did like this:
https://github.com/terlar/nix-config/blob/master/lib/scripts.nix#L11-L35

I experimented by putting the full logic inside the nix eval, but found that was a bit harder to read, so opted for a bash function that could do the check and be used multiple times.

Reply:
I have added a modified version of your implementation. I tried to simplify it a bit, but didn't change much overall.
This version will only try to add the hostname if you don't give it a name, meaning `--flake uri#name` won't try to build name@hostname, otherwise it might be impossible to switch to certain attributes (if you have name *and* name@hostname).